### PR TITLE
Enable clang warnings used by LLVM

### PR DESCRIPTION
**Pull request by klausler**
Enable `-Wstring-conversion` and `-Wcovered-switch-default` warnings when building with clang.  Patch sources to dodge the warnings.

### CODE REVIEW

File: lib/common/idioms.h [hunk 1]
```diff
@@ -74,6 +74,16 @@ template<typename... LAMBDAS> visitors(LAMBDAS... x)->visitors<LAMBDAS...>;
 // For switch statements without default: labels.
 #define CRASH_NO_CASE DIE("no case")
 
+// For switch statements whose cases have return statements for
+// all possibilities.  Clang emits warnings if the default: is
+// present, gcc emits warings if it is absent.
```
Comment (tskeith) [inline lib/common/idioms.h]: warings -> warnings

File: lib/common/idioms.h [hunk 2]
```diff
@@ -74,6 +74,16 @@ template<typename... LAMBDAS> visitors(LAMBDAS... x)->visitors<LAMBDAS...>;
...
+#if __clang__
+#define SWITCH_COVERS_ALL_CASES
+#else
+#define SWITCH_COVERS_ALL_CASES \
+  default: CRASH_NO_CASE; \
+           #endif
```
Comment (tskeith) [inline lib/common/idioms.h]: The backslash on the previous line shouldn't be there.

File: lib/parser/characters.cc [hunk 3]
```diff
@@ -122,8 +122,9 @@ template<> EncodedCharacter EncodeCharacter<Encoding::UTF_8>(char32_t ucs) {
 EncodedCharacter EncodeCharacter(Encoding encoding, char32_t ucs) {
   switch (encoding) {
   case Encoding::LATIN_1: return EncodeCharacter<Encoding::LATIN_1>(ucs);
-  case Encoding::UTF_8: return EncodeCharacter<Encoding::UTF_8>(ucs);
-  default: CRASH_NO_CASE;
+  case Encoding::UTF_8:
+    return EncodeCharacter<Encoding::UTF_8>(ucs);
+    SWITCH_COVERS_ALL_CASES
```
Comment (tskeith) [inline lib/parser/characters.cc]: clang-format treates `SWITCH_COVERS_ALL_CASES` like it is part of the last case.
What do you think of putting it between the `switch` and the first `case` to make it clear it is separate?
```
  switch (encoding) {
    SWITCH_COVERS_ALL_CASES
  case Encoding::LATIN_1: return EncodeCharacter<Encoding::LATIN_1>(ucs);
  case Encoding::UTF_8: return EncodeCharacter<Encoding::UTF_8>(ucs);
  }
```

File: lib/parser/prescan.cc [hunk 4]
```diff
@@ -58,7 +58,7 @@ static void NormalizeCompilerDirectiveCommentMarker(TokenSequence &dir) {
       return;
     }
   }
-  CHECK(!"compiler directive all blank");
+  CHECK(false && "compiler directive all blank");
```
Comment (tskeith) [inline lib/parser/prescan.cc]: Could use `DIE` instead.

File: lib/semantics/resolve-names.cc [hunk 5]
```diff
@@ -4824,8 +4824,7 @@ const DeclTypeSpec &ConstructVisitor::ToDeclTypeSpec(
 
       );
     }
-  case common::TypeCategory::Character:
-  default: CRASH_NO_CASE;
+  case common::TypeCategory::Character: CRASH_NO_CASE; SWITCH_COVERS_ALL_CASES
```
Comment (tskeith) [inline lib/semantics/resolve-names.cc]: The clang-format problem is even worse here.

File: lib/common/idioms.h [hunk 6]
```diff
@@ -74,6 +74,15 @@ template<typename... LAMBDAS> visitors(LAMBDAS... x)->visitors<LAMBDAS...>;
 // For switch statements without default: labels.
 #define CRASH_NO_CASE DIE("no case")
 
+// For switch statements whose cases have return statements for
+// all possibilities.  Clang emits warnings if the default: is
+// present, gcc emits warnings if it is absent.
+#if __clang__
```
Comment (sscalpone) [inline lib/common/idioms.h]: Please update the style guide.

File: lib/semantics/check-omp-structure.cc [hunk 7]
```diff
@@ -585,9 +585,6 @@ void OmpStructureChecker::Enter(
     OmpClauseSet allowed{OmpClause::DEPEND};
     SetContextAllowed(allowed);
   } break;
-  default:
-    // TODO others
-    break;
```
Comment (DavidTruby) [inline lib/semantics/check-omp-structure.cc]: This switch statement should also have `SWITCH_COVERS_ALL_CASES` as per the updated style guide

Comment (RichBarton-Arm) [reply lib/semantics/check-omp-structure.cc]: @klausler looks like @DavidTruby's comment fell through the cracks. @DavidTruby - can you submit a separate PR for it?

**Pull request merged**